src/action.ts

```diff
@@ -123,40 +124,86 @@ export async function performAction(
 	tracker: string
 ): Promise<ActionResult> {
 	const { action, linkDir } = getRuntimeConfig();
-	let linkedFilesRoot: string | undefined;
 
-	if (linkDir) {
-		const linkedFilesRootResult = await linkAllFilesInMetafile(
+	// SAVES FIRST, PREVENTS LINKING THOUGH IN THIS ORDER
+	if (action === Action.SAVE) {
+		saveTorrentFile(tracker, getTag(searchee.name), newMeta);
+		const styledName = chalk.green.bold(newMeta.name);
+		const styledTracker = chalk.bold(tracker);
+		logger.info(`Found ${styledName} on ${styledTracker} - saved`);
+		return SaveResult.SAVED;
+	}
+
+	let destinationDir: string | undefined,
+		ogDownloadDir,
+		linkedFilesRootResult;
+
+	// SET THE CLIENTS SAVE PATH - PREVENTS ERROR IF DATADIR/LINKDIR IS SET
+	// BUT NOT THE TORRENT DATA DIRECTORY AND/OR LINKING THROWS OR IS NOT SET
+	if (searchee.infoHash) {
+		ogDownloadDir = await getClient().getDownloadDir(searchee);
+		destinationDir = ogDownloadDir.isOk()
+			? ogDownloadDir.unwrapOrThrow()
+			: undefined;
+	}
+	if (
+		linkDir &&
+		typeof destinationDir === "string" &&
+		existsSync(destinationDir)
+	) {
+		linkedFilesRootResult = await linkAllFilesInMetafile(
 			searchee,
 			newMeta,
 			tracker,
 			decision
 		);
-		if (linkedFilesRootResult.isErr()) {
-			// TODO
-			return InjectionResult.FAILURE;
-		}
-		linkedFilesRoot = linkedFilesRootResult.unwrapOrThrow();
+		destinationDir = linkedFilesRootResult.isOk()
+			? linkedFilesRootResult.unwrapOrThrow()
+			: destinationDir;
+		/*} else {
+			//linking failed - need to compare file trees or figure out nested linking?
+			//compareFileTrees(newMeta, searchee);
+			//	logInjectionResult(result, tracker, newMeta.name);
+
+			destinationDir = ogDownloadDir.unwrapOrThrow();
+
+			/* extra code for referencing
+				const result =
+					linkedFilesRootResult.unwrapErrOrThrow() as InjectionResult;
+				logInjectionResult(result, tracker, newMeta.name);
+				saveTorrentFile(tracker, getTag(searchee.name), newMeta);
+				return InjectionResult.FAILURE;
+				*/
+	}
+	if (typeof destinationDir !== "string") {
+		logInjectionResult(InjectionResult.FAILURE, tracker, newMeta.name);
+		saveTorrentFile(tracker, getTag(searchee.name), newMeta);
+		return InjectionResult.FAILURE;
 	}
 
-	if (action === Action.INJECT) {
-		// using downloadDir for now but we might want to use linkedFilesRoot instead if we get missing files errors
-		const downloadDir = linkedFilesRoot
-			? dirname(linkedFilesRoot)
-			: linkedFilesRoot;
-		const result = await getClient().inject(newMeta, searchee, downloadDir);
-		logInjectionResult(result, tracker, newMeta.name);
-		if (result === InjectionResult.FAILURE) {
-			saveTorrentFile(tracker, getTag(searchee.name), newMeta);
-		}
-		return result;
-	} else {
+	// NEW compare() for STRICTLY COMPARING FILES (nested linking)
+	const nestedMatch = compareFileTreesIgnoringFolders(newMeta, searchee);
+	const perfectMatch = compareFileTrees(newMeta, searchee);
+
```

Comment:
I think it'd be good to keep the decision code separate from the action code. 
Could we achieve this by checking that the decision was `MATCH` vs `MATCH_SIZE_ONLY`? or maybe do we need to add another member to that `Decision` enum to represent this? 

```suggestion
	// NEW compare() for STRICTLY COMPARING FILES (nested linking)
	const isNestedMatch = compareFileTreesIgnoringFolders(newMeta, searchee);
	const isPerfectMatch = compareFileTrees(newMeta, searchee);

```

Reply:
I think something we may run into with that approach without some changes to the `Decision` return is that we could have multi file torrents match on size that are completely different. This basically will determine if they are the same torrents but nested, or perfect matches.